File: transfermarkt/src/main/java/com/liordahan/mgsrteam/transfermarket/PlayersUpdate.kt
package com.liordahan.mgsrteam.transfermarket

import android.net.Network
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.IOException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

data class PlayerToUpdateValues(
    val marketValue: String?,
    val profileImage: String?,
    val nationalityFlag: String?,
    val citizenship: String?,
    val citizenships: List<String> = emptyList(),
    val citizenshipFlags: List<String> = emptyList(),
    val age: String?,
    val contract: String?,
    val positions: List<String?>?,
    val currentClub: TransfermarktClub?,
    val isOnLoan: Boolean = false,
    val onLoanFromClub: String? = null,
    val foot: String? = null,
    val agency: String? = null,
    val agencyUrl: String? = null
)

class PlayersUpdate {

    /**
     * @param network optional [Network] to route the request through (for IP rotation).
     *                When non-null the request is opened via [Network.openConnection] so it
     *                travels over that specific interface (WiFi / Cellular).
     */
    suspend fun updatePlayerByTmProfile(
        tmProfile: String?,
        network: Network? = null
    ): TransfermarktResult<PlayerToUpdateValues?> =
        withContext(Dispatchers.IO) {
            val profileUrl = tmProfile?.trim()
                ?: return@withContext TransfermarktResult.Failed("Profile URL is null or blank")

            // Primary: web proxy API (bypasses Cloudflare TLS fingerprinting)
            if (network == null) {
                try {
                    val result = updatePlayerViaProxy(profileUrl)
                    if (result != null) return@withContext TransfermarktResult.Success(result)
                } catch (_: Exception) { /* fall through to direct scraping */ }
            }

            // Fallback: direct scraping (or when specific network is requested for IP rotation)
            return@withContext try {
                val doc = if (network != null) {
                    TransfermarktHttp.fetchDocument(profileUrl, network)
                } else {
                    TransfermarktHttp.fetchDocument(profileUrl)
                }

                val (citizenships, citizenshipFlags) = extractAllNationalitiesFromProfile(doc)
                val citizenship = citizenships.firstOrNull().orEmpty()
                val flag = citizenshipFlags.firstOrNull().orEmpty()

                val contract = doc.select("span.data-header__label")
                    .text()
                    .substringAfterLast(":")
                    .trim()

                val playerImage = doc
                    .select("div.data-header__profile-container img")
                    .firstOrNull()
                    ?.attr("src")
                    .orEmpty()

                val marketValue = doc.select("div.data-header__box--small")
                    .text()
                    .substringBefore("Last")
                    .trim()

                val positionsList = doc.select("div.detail-position__box dd")
                    .mapNotNull {
                        it.text()
                            .replace("-", " ")
                            .takeIf(String::isNotBlank)
                            ?.convertLongPositionNameToShort()
                    }
                    .ifEmpty {
                        val fallback = doc
                            .select("div.data-header__info-box ul.data-header__items")
                            .getOrNull(1)
                            ?.text()
                            ?.substringAfterLast(":")
                            ?.trim()
                        listOfNotNull(fallback)
                    }

                val age = doc.select("span[itemprop=birthDate]")
                    .firstOrNull()
                    ?.text()
                    ?.substringAfter("(")
                    ?.substringBefore(")")
                    .orEmpty()

                val clubName = doc.select("span.data-header__club")
                    .select("a")
                    .attr("title")

                val clubLogo = doc.select("div.data-header__box--big")
                    .select("img")
                    .attr("srcset")
                    .substringBefore("1x")
                    .trim()

                val clubTmProfile = TRANSFERMARKT_BASE_URL +
                        doc.select("span.data-header__club")
                            .select("a")
                            .attr("href")

                val clubCountry = doc
                    .select("div.data-header__club-info")
                    .select("span.data-header__label")
                    .select("img")
                    .attr("title")

                val club = TransfermarktClub(
                    clubName = clubName,
                    clubLogo = clubLogo,
                    clubTmProfile = clubTmProfile,
                    clubCountry = clubCountry
                )

                val loanInfo = detectLoanStatus(doc, clubName)

                val foot = extractFootFromDocument(doc, null)

                val infoLabels = doc.select("span.info-table__content--regular")
                var agency: String? = null
                var agencyUrl: String? = null

                for (label in infoLabels) {
                    val labelText = label.text().trim().lowercase()
                    val valueSpan = label.nextElementSibling() ?: continue

                    when {
                        labelText.contains("player agent") || labelText.contains("agent") -> {
                            val link = valueSpan.selectFirst("a")
                            agency = link?.text()?.trim()?.takeIf { it.isNotBlank() }
                                ?: valueSpan.text().trim().takeIf { it.isNotBlank() }
                            val href = link?.attr("href")
                            if (!href.isNullOrBlank()) {
                                agencyUrl = if (href.startsWith("http")) href
                                else TRANSFERMARKT_BASE_URL + href
                            }
                        }
                    }
                }

                TransfermarktResult.Success(
                    PlayerToUpdateValues(
                        marketValue = marketValue,
                        profileImage = playerImage,
                        nationalityFlag = flag,
                        citizenship = citizenship,
                        citizenships = citizenships,
                        citizenshipFlags = citizenshipFlags,
                        age = age,
                        contract = contract,
                        positions = positionsList,
                        currentClub = club,
                        isOnLoan = loanInfo.isOnLoan,
                        onLoanFromClub = loanInfo.onLoanFromClub,
                        foot = foot,
                        agency = agency,
                        agencyUrl = agencyUrl
                    )
                )
            } catch (ex: Exception) {
                TransfermarktResult.Failed(ex.localizedMessage ?: "Unknown error")
            }
        }

    /** Fetch player update data via the Next.js web proxy API. */
    private fun updatePlayerViaProxy(profileUrl: String): PlayerToUpdateValues? {
        val encoded = URLEncoder.encode(profileUrl, StandardCharsets.UTF_8.toString())
        val url = "$WEB_PROXY_BASE/api/transfermarkt/player?url=$encoded"
        val json = TransfermarktHttp.fetchStringSync(url)
        val p = JSONObject(json)
        if (p.has("error")) throw IOException(p.optString("error", "Proxy error"))
        val fullName = p.optString("fullName", "")
        if (fullName.isBlank()) return null // proxy returned no useful data

        val nationalities = mutableListOf<String>()
        val nationalityFlags = mutableListOf<String>()
        p.optJSONArray("nationalities")?.let { arr ->
            for (i in 0 until arr.length()) nationalities.add(arr.getString(i))
        }
        p.optJSONArray("nationalityFlags")?.let { arr ->
            for (i in 0 until arr.length()) nationalityFlags.add(arr.getString(i))
        }
        val positions = mutableListOf<String>()
        p.optJSONArray("positions")?.let { arr ->
            for (i in 0 until arr.length()) positions.add(arr.getString(i))
        }
        val clubObj = p.optJSONObject("currentClub")

        return PlayerToUpdateValues(
            marketValue = p.optString("marketValue", null)?.takeIf { it.isNotBlank() },
            profileImage = p.optString("profileImage", null)?.takeIf { it.isNotBlank() },
            nationalityFlag = nationalityFlags.firstOrNull(),
            citizenship = nationalities.firstOrNull(),
            citizenships = nationalities,
            citizenshipFlags = nationalityFlags,
            age = p.optString("age", null)?.takeIf { it.isNotBlank() },
            contract = p.optString("contractExpires", null)?.takeIf { it.isNotBlank() },
            positions = positions.ifEmpty { null },
            currentClub = clubObj?.let {
                TransfermarktClub(
                    clubName = it.optString("clubName", null),
                    clubLogo = it.optString("clubLogo", null),
                    clubTmProfile = it.optString("clubTmProfile", null),
                    clubCountry = it.optString("clubCountry", null),
                )
            },
            isOnLoan = p.optBoolean("isOnLoan", false),
            onLoanFromClub = p.optString("onLoanFromClub", null)?.takeIf { it != "null" && it.isNotBlank() },
            foot = p.optString("foot", null)?.takeIf { it.isNotBlank() },
            agency = p.optString("agency", null)?.takeIf { it != "null" && it.isNotBlank() },
            agencyUrl = p.optString("agencyUrl", null)?.takeIf { it != "null" && it.isNotBlank() },
        )
    }
}

File: workers-job/lib/playersUpdate.js
/**
 * PlayersUpdate — Node.js port of Kotlin PlayersUpdate.
 * Fetches a Transfermarkt player profile and extracts update values.
 */

const axios = require("axios");
const cheerio = require("cheerio");
const { TRANSFERMARKT_BASE_URL } = require("./utils");

const USER_AGENTS = [
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/122.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/122.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0",
];

function getRandomUserAgent() {
  return USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)];
}

async function fetchDocument(url) {
  const res = await axios.get(url, {
    headers: {
      "User-Agent": getRandomUserAgent(),
      "Accept-Language": "en-US,en;q=0.9",
    },
    timeout: 12000,
    validateStatus: (s) => s === 200,
  });
  return cheerio.load(res.data);
}

const LOAN_FROM_UNTIL = /(?:on loan from|leihe von|ausgeliehen von)\s*:?\s*(.+?)\s+(?:contract|until|bis)/i;
const LOAN_FROM_EOL = /(?:on loan from|leihe von|ausgeliehen von)\s*:?\s*(.+?)(?:\s*$|\s*;)/i;

function detectLoanStatus($, clubName) {
  const ribbon =
    $("div.data-header_ribbon, div.data-header__ribbon").first()[0] ||
    $("div[class*='ribbon']").first()[0];
  const ribbonLinkTitleRaw = ribbon
    ? $(ribbon).find("a").first().attr("title") || ""
    : $("a[title*='on loan from']").first().attr("title") || "";
  const ribbonText = ribbon ? $(ribbon).text().trim().toLowerCase() : "";
  const clubSectionText = $("span.data-header__club, div.data-header__club-info").text().toLowerCase();
  const infoBoxText = $("div.data-header__info-box").text().toLowerCase();
  const headerText = $("div.data-header").text().toLowerCase();
  const combined = `${ribbonLinkTitleRaw} ${ribbonText} ${clubSectionText} ${infoBoxText} ${headerText}`.toLowerCase();

  const hasLoanIndicator =
    ribbonLinkTitleRaw.toLowerCase().includes("on loan from") ||
    combined.includes("on loan") ||
    combined.includes("leihe") ||
    combined.includes("ausgeliehen") ||
    combined.includes("leihe von") ||
    combined.includes("ausgeliehen von") ||
    combined.includes("prêt") ||
    combined.includes("en préstamo") ||
    combined.includes("in prestito") ||
    (combined.includes("loan") &&
      !combined.includes("end of loan") &&
      !combined.includes("loan return") &&
      !combined.includes("loan spell"));
  const isReturnee =
    combined.includes("returnee") || combined.includes("returned after loan");
  const isOnLoan = hasLoanIndicator && !isReturnee;

  let onLoanFromClub = null;
  if (isOnLoan) {
    const headerTextRaw = $("div.data-header").text();
    const infoBoxTextRaw = $("div.data-header__info-box").text();
    const searchText = ribbonLinkTitleRaw || headerTextRaw || infoBoxTextRaw;
    const m1 = LOAN_FROM_UNTIL.exec(searchText);
    const m2 = LOAN_FROM_EOL.exec(searchText);
    onLoanFromClub =
      (m1 && m1[1]?.trim()) ||
      (m2 && m2[1]?.trim()) ||
      null;
    if (!onLoanFromClub) {
      const vereinLinks = $("div.data-header a[href*='/verein/']");
      vereinLinks.each((i, el) => {
        const $el = $(el);
        const title = $el.attr("title")?.trim();
        const text = $el.text().trim();
        const val = title || text;
        if (val && val !== clubName) onLoanFromClub = val;
        return !onLoanFromClub;
      });
    }
  }

  return { isOnLoan, onLoanFromClub };
}

const FOOT_REGEX = /(?:Foot|Fuss|Preferred\s+foot)\s*:?\s*(\w+)/i;
const VALID_FOOT = new Set(["left", "right", "both", "links", "rechts", "beide", "l", "r", "b"]);

function extractFoot($) {
  let result = null;
  $("span.info-table__content--regular").each((i, el) => {
    const label = $(el).text().trim().toLowerCase();
    if (label.includes("foot") || label.includes("fuss")) {
      const next = $(el).next();
      const val = next.text().trim().toLowerCase();
      if (val && (VALID_FOOT.has(val) || val.startsWith("left") || val.startsWith("right"))) {
        if (val.startsWith("left") || val === "l") result = "Left";
        else if (val.startsWith("right") || val === "r") result = "Right";
        else if (val.includes("both") || val === "b") result = "Both";
        else result = val;
        return false;
      }
    }
  });
  if (result) return result;
  const bodyText = $("body").text();
  const m = FOOT_REGEX.exec(bodyText);
  if (m && m[1]) {
    const v = m[1].toLowerCase();
    if (VALID_FOOT.has(v) || v.startsWith("left") || v.startsWith("right")) {
      if (v.startsWith("left")) return "Left";
      if (v.startsWith("right")) return "Right";
      return "Both";
    }
  }
  return null;
}

function convertPosition(pos) {
  const map = {
    Goalkeeper: "GK", "Left Back": "LB", "Centre Back": "CB", "Right Back": "RB",
    "Defensive Midfield": "DM", "Central Midfield": "CM", "Attacking Midfield": "AM",
    "Right Winger": "RW", "Left Winger": "LW", "Centre Forward": "CF",
    "Second Striker": "SS", "Left Midfield": "LM", "Right Midfield": "RM",
  };
  return map[pos] || pos || "";
}

async function updatePlayerByTmProfile(tmProfile) {
  const url = (tmProfile || "").trim();
  if (!url) return { success: false, error: "Profile URL is null or blank" };

  try {
    const $ = await fetchDocument(url);

    // Info-table Citizenship row has ALL citizenships; header itemprop only has primary
    const citizenshipLabel = $('span.info-table__content--regular').filter(function() {
      return $(this).text().trim().startsWith('Citizenship');
    });
    const citizenshipContent = citizenshipLabel.next('.info-table__content--bold');
    let natEls = citizenshipContent.find('img');
    if (!natEls.length) natEls = $("[itemprop=nationality] img");
    const citizenships = [];
    const citizenshipFlags = [];
    natEls.each((_, el) => {
      const title = $(el).attr("title");
      if (title) citizenships.push(title);
      let src = $(el).attr("src") || "";
      if (src) citizenshipFlags.push(src.replace("tiny", "head").replace("verysmall", "head"));
    });
    const citizenship = citizenships[0] || "";
    let flag = citizenshipFlags[0] || "";

    const contractText = $("span.data-header__label").text();
    const contract = contractText.split(":").pop()?.trim() || "";

    const playerImage =
      $("div.data-header__profile-container img").first().attr("src") || "";

    const marketValueBox = $("div.data-header__box--small").text();
    const marketValue = marketValueBox.split("Last")[0].trim();

    const positionsList = [];
    $("div.detail-position__box dd").each((i, el) => {
      const t = $(el).text().replace(/-/g, " ").trim();
      if (t) positionsList.push(convertPosition(t));
    });
    if (positionsList.length === 0) {
      const fallback = $("div.data-header__info-box ul.data-header__items").eq(1).text();
      const p = fallback.split(":").pop()?.trim();
      if (p) positionsList.push(convertPosition(p));
    }

    const birthEl = $("span[itemprop=birthDate]").first().text();
    const age = birthEl ? birthEl.replace(/.*\(([^)]+)\).*/, "$1") : "";

    const clubLink = $("span.data-header__club a");
    const clubName = clubLink.attr("title") || "";
    const clubLogoEl = $("div.data-header__box--big img").first().attr("srcset") || "";
    const clubLogo = clubLogoEl.split("1x")[0].trim();
    const clubHref = clubLink.attr("href") || "";
    const clubTmProfile = clubHref.startsWith("http") ? clubHref : TRANSFERMARKT_BASE_URL + clubHref;
    const clubCountry = $("div.data-header__club-info span.data-header__label img").attr("title") || "";

    const loanInfo = detectLoanStatus($, clubName);
    const foot = extractFoot($);

    let agency = null;
    let agencyUrl = null;
    $("span.info-table__content--regular").each((i, el) => {
      const label = $(el).text().trim().toLowerCase();
      if (label.includes("player agent") || label.includes("agent")) {
        const next = $(el).next();
        const link = next.find("a").first();
        agency = link.text().trim() || next.text().trim();
        const href = link.attr("href");
        if (href) {
          agencyUrl = href.startsWith("http") ? href : TRANSFERMARKT_BASE_URL + href;
        }
      }
    });

    return {
      success: true,
      data: {
        marketValue,
        profileImage: playerImage,
        nationalityFlag: flag,
        citizenship,
        citizenships,
        citizenshipFlags,
        age,
        contract,
        positions: positionsList,
        currentClub: {
          clubName,
          clubLogo,
          clubTmProfile,
          clubCountry,
        },
        isOnLoan: loanInfo.isOnLoan,
        onLoanFromClub: loanInfo.onLoanFromClub,
        foot,
        agency,
        agencyUrl,
      },
    };
  } catch (err) {
    return {
      success: false,
      error: err.message || String(err),
    };
  }
}

module.exports = { updatePlayerByTmProfile };
